### ISSUE TITLE
Fix flaky tests

### ISSUE THREAD

**Issue by ericdallo** [clojure-lsp/clojure-lsp#704]
We already had flaky tests in the past and a lot of them were already fixed, but we still have some flaky tests especially on integration tests.

**Comment by ericdallo** [comment]
It seems to me that the integration-test `diagnostics_test` is the most (only?) flaky ATM

**Comment by mainej** [comment]
I looked at this a bit, but didn't make much progress. Here's what I found...

I see two tests that fail intermittently. 
1. `integration-test/integration/settings_change_test.clj:16` [Ex](https://github.com/mainej/clojure-lsp/runs/4907925973#step:7:107)
2. `integration-test/integration/diagnostics_test.clj:10` [Ex](https://github.com/clojure-lsp/clojure-lsp/runs/4899541082#step:7:71)

Both tests analyze `sample-test/src/sample_test/diagnostics/unused_public_var.clj`, and in both cases they are expecting diagnostics that report two unused public vars for that file. When they fail, they see no diagnostics.

I've identified a few possible sources of flakiness, though I'm not sure this is all of them:
1. In failing tests, the `await-diagnostics` loop could be finding the wrong diagnostics message. I don't think this is the case because the test logs seem to show that the same messages are sent and received whether the test passes or fails. When the test fails a few later messages aren't sent, but I assume that's because the tests are in some sort of fail-fast mode.
2. In failing tests, the `.lsp/config.edn` could be stale, with the public-var linting turned off.
3. In failing tests, the kondo cache could be used instead of a fresh run, or vice versa.
4. In failing tests, the kondo cache could be stale, or be a mismatch with what you'd expect for the current `.lsp/config.edn`.
5. In failing tests, the kondo analysis could fail to detect unused public vars.
6. In failing tests, the kondo analysis could be running asynchronously. It _appears_ that clojure-lsp invokes kondo synchronously (i.e. not `:parallel`) for public var linting but it may be worth researching.
7. In failing tests, the :project-root could be wrong. If this is happens, the linted file is considered "external" to the project, and is therefore excluded from public var linting. This doesn't seem to be the case, because the `initialize-request` fixture sets it in both passing and failing runs.

I don't have the time to keep looking at this right now, but here are some avenues of further research:

1. Confirm whether the tests are in a fail-fast mode. If not, research why message logs are different in passing and failing tests.
2. Log the contents of`.lsp/config.edn` before each test, to ensure public var linting is on/off as expected.
3. Wrap the re-write of `.lsp/config.edn` in `settings_change_test` in a `(try ... (finally ...))`. This seems prudent regardless, and especially important if the tests are in a fail-fast mode, which would mean the config file isn't always restored.
4. Log whether the diagnostics are based on the kondo cache or on a fresh run (how?)
5. Log whether the kondo cache contains any diagnostics for the file (not just "project internal" diagnostics).
6. Research whether kondo analysis is running, but failing to detect unused public vars (how?)
7. Research whether kondo analysis is running asynchronously. (how?)

**Comment by ericdallo** [comment]
Thanks for the findings, all of that makes sense to me and I already spend some time researching the same, glad you found the same I found at the time, all those ideas sounds good to start debugging that, thanks for that!

> 3. Wrap the re-write of .lsp/config.edn in settings_change_test in a (try ... (finally ...)). This seems prudent regardless, and especially important if the tests are in a fail-fast mode, which would mean the config file isn't always restored.

That sounds like a good idea

**Comment by mainej** [comment]
Here's a [variation](https://github.com/clojure-lsp/clojure-lsp/runs/4910705932?check_suite_focus=true#step:7:115) I haven't seen before. Kondo reports one unused public var. Weird. That makes me think it's a problem in kondo. I guess it could be caching, but still... interesting.

**Comment by ericdallo** [comment]
Yeah, it even seems it's becoming more frequent :(

**Comment by mainej** [comment]
@ericdallo I think I found the culprit (or one of the culprits anyway.) See #724

**Comment by ericdallo** [comment]
Amazing finding!

**Comment by ericdallo** [comment]
I feel the recent changes improved a lot the flakiness of some tests!

**Comment by mainej** [comment]
Agreed, the build failures have been much less frequent. :)

When it does fail, it's usually a timeout waiting for a request, response or notification. Fundamentally, there must be some problem in the code—a race condition, a crash, or something else—which means the message is never seen. But it's pretty hard to debug with the amount of information we have. Is there some way to get access to the log files when the test suite times out?

Also, [that code](https://github.com/clojure-lsp/clojure-lsp/blob/master/cli/integration-test/integration/lsp.clj#L104-L171) has no limit to the number of times it'll loop. Would it be better if it failed after some number of tries? The suite would still be marked as a failure, but it would at least finish a little faster.

**Comment by ericdallo** [comment]
> But it's pretty hard to debug with the amount of information we have. Is there some way to get access to the log files when the test suite times out?

No we don't have, but we can try two things:
- make clojure-lsp on CI log to stderr instead of custom log file (check clojure-lsp.logging)
- When test fail or something we just cat/tail the log file? that seems easier we just need to pass `:log-path` setting on initializationOptions of the test and cat it later


> Also, [that code](https://github.com/clojure-lsp/clojure-lsp/blob/master/cli/integration-test/integration/lsp.clj#L104-L171) has no limit to the number of times it'll loop. Would it be better if it failed after some number of tries? The suite would still be marked as a failure, but it would at least finish a little faster.

Yeah, but I have no idea how much retries makes sense, maybe we can add a flag to log retries on CI to learn more about it.

**Comment by mainej** [comment]
> When test fail or something we just cat/tail the log file

That's a nice idea. Logging to stderr would be OK too, but would add a lot of noise for successful tests. Also there's already a lot of magic happening with `*stdout*`. It might get confusing to add `*stderr*` into the mix.

> add a flag to log retries on CI to learn more about it

You mean log how many tries it takes in passing tests, so we know what a reasonable limit would be? That could be interesting. I bet it's not very many tries. When I've watched the test runs, I haven't noticed long pauses. There's a 500ms wait between attempts, which makes me think it usually succeeds within the first few tries.

**Comment by ericdallo** [comment]
exactly

**Comment by ericdallo** [comment]
@mainej I noticed we async lint the file when `didOpen` happens, which can explain the diagnostics not being available on some integration tests?

**Comment by mainej** [comment]
@ericdallo that's interesting. I haven't seen any flaky tests recently about diagnostics. I think those were fixed by #724. But more importantly, it means that sometimes analysis is missing too. I guess what's surprising is that the tests pass so often, even when analysis hasn't completed. I guess one way to understand which tests might be affected is to comment out the merging of the analysis. I'll try that and let you know what I find.

If that's the problem, what do you think we should do? We could wait for diagnostics after opening every file. But as a user, diagnostics are not what I notice, because most files don't have any diagnostics. Instead, I wait until semantic tokens are available. That seems to be a good signal that everything is ready. Maybe the tests should do the same? 

Also, I did some analysis of the flaky failures since #724. I only see 4:

1. https://github.com/clojure-lsp/clojure-lsp/runs/5025824881?check_suite_focus=true
2. https://github.com/clojure-lsp/clojure-lsp/runs/4988722765?check_suite_focus=true
3. https://github.com/clojure-lsp/clojure-lsp/runs/4988454455?check_suite_focus=true
4. https://github.com/clojure-lsp/clojure-lsp/runs/4961842514?check_suite_focus=true

None of these failed because they get the _wrong_ results. They failed because they get _no_ results. The server never replied, or replied, but never sent follow-up requests. To me that suggests that the server is throwing exceptions. (And perhaps those exceptions are because the code expects analysis to be finished, in-line with your observation about `didOpen` being async.)

In 1, the client invoked a command and expected the server to ask for two changes—a text edit and a cursor reposition. The text edit happened but the cursor reposition didn't.

In 2-4, the test timed out waiting for a response from the server. (Look for a line "Sending request:..." with no corresponding "Received response:...".) In number 2 and 4, the client requested textDocument/linkedEditingRange. In 3, the client requested textDocument/formatting. Those aren't async, as far as I can see. I'll check whether they throw exceptions if they're invoked before analysis finishes.

**Comment by ericdallo** [comment]
Makes sense to me @mainej, we really need a way to get server logs on integration tests to confirm that, it looks promising that analysis, thank you!

**Comment by mainej** [comment]
Hmm... I'm stumped again. Looking at it more carefully, `didOpen` _synchronously_ lints the file, but then _asynchronously_ publishes diagnostics. The tests wait for a response from `didOpen`, so, analysis and diagnostics will always be available on the server, even though the client may not know about the diagnostics yet.

Niether textDocument/linkedEditingRange nor textDocument/formatting throw errors when analysis is missing, so that theory is wrong.

I haven't tried getting server logs into the integration tests, but I agree, that's going to be our best chance of figuring this out.

**Comment by ericdallo** [comment]
> Hmm... I'm stumped again. Looking at it more carefully, didOpen synchronously lints the file, but then asynchronously publishes diagnostics. The tests wait for a response from didOpen, so, analysis and diagnostics will always be available on the server, even though the client may not know about the diagnostics yet.

You are right, only the publish to client is async.

**Comment by mainej** [comment]
See https://github.com/clojure-lsp/clojure-lsp/runs/5150490400?check_suite_focus=true for the first failure which includes logs. I don't have time to investigate it right now, but will tomorrow.

**Comment by mainej** [comment]
@ericdallo why does the client loop swallow java.io.IOException errors? https://github.com/clojure-lsp/clojure-lsp/blob/ca1621ba7f8475095f2bf74d895411372debf4f4/cli/integration-test/integration/lsp.clj#L69

Also, to summarize the analysis of the logs after #757... the server logs suggest it's doing everything right, sending the responses it should. Sometimes, however, the client fails to receive a response. That's made us think the problem is outside of the server: in the client-side test harness code, or perhaps in lsp4j.

**Comment by ericdallo** [comment]
IMHO it's probably the bb client side loop, I can't recall why I added that catch on IOException, but sounds like a good start 😅

**Comment by mainej** [comment]
I had an integration test timeout locally. I ran it again and it passed, so I think it was the real deal—the same as we see on CI.

I was able to grab a stack dump with `jstack` while it was frozen. I haven't had time to dig into it much, but what stands out initially is that there are two threads parked in lsp4j code. That may be a common situation, or it may indicate some sort of deadlock. I have practically no experience with Java stack dumps or debugging deadlocks, so I'd appreciate hearing anyone else's interpretations.

I'm also attaching the server logs... It appears that the server thinks it sent a message ([here](https://github.com/clojure-lsp/clojure-lsp/blob/aef068ab2550be1b17f85ae7952f7e153cfa0565/cli/integration-test/integration/code_action_test.clj#L73)), but the client never received it.

[stack_dump.txt](https://github.com/clojure-lsp/clojure-lsp/files/8456011/stack_dump.txt)
[test_logs.txt](https://github.com/clojure-lsp/clojure-lsp/files/8456012/test_logs.txt)

**Comment by ericdallo** [comment]
Good observation, @mainej, did you confirm server really sent that message? or client didn't catch properly on client side?

**Comment by mainej** [comment]
> did you confirm server really sent that message? or client didn't catch properly on client side?

I'm not sure how to do that. I don't think the messages get logged anywhere. Could that be what `lsp4clj.core/tee-system-out` is for? As a debugging tool for these kinds of problems? (If not, what _is_ it for? It doesn't look like it's ever actually called.)

What I can see is that clojure-lsp (or really lsp4clj) printed a line right before it returned control to lsp4j, [here](https://github.com/clojure-lsp/lsp4clj/blob/907d73ba5174c470eb672ba6b36fdc6734b68c8f/server/src/lsp4clj/core.clj#L375-L379).

Here's where I get hazy... Then lsp4j is supposed to print the response to `System/out`. Is that right? I think that's how we have it configured. 

Next, we have a "client" in `integration.lsp`. It's not a client in the lsp4j sense, it's just a loop that reads the server's output. The "client" parses the output, classifies it by message type, saves it in one of a few atoms, and prints it.

Our test code waits for the "client" to put a particular message into the atom, which in this case was [here](https://github.com/clojure-lsp/clojure-lsp/blob/1d6e554ec0afd3a04c113e513f6b6724bbe0f709/cli/integration-test/integration/code_action_test.clj#L73). The "client" never printed the message or put it in the atom.

So, all I can really say is that something went bad between the server printing that line, and the integration test picking the response out of the atom, which is as much information as we had before.

The problem could be in the last few steps lsp4clj does before sending the message, or it could be in our "client" loop. But we've done a lot to rule those out. We're still not 100% sure that the problem is in lsp4j code, but if it is, I'd start with a few theories.

First, it's possible that using `System/out` is a bad idea in tests. `System/out` is a _buffered_ OutputStream. As I understand it, this means that if lsp4j doesn't `flush` that stream after writing its message, there are no guarantees that consumers of that stream will ever get the message.

In the real world (i.e., not in tests) the client keeps sending messages, and the server keeps replying, which means that buffer always fills up and gets automatically flushed. So, we wouldn't see this problem in the real world. I'll plan to look around in the lsp4j code for `flush`es. I find that code hard to follow, so even if I find a `flush`, I might not be able to tell whether our messages are being `flushed`.

Second, it's possible that lsp4j sometimes drops its responses. I don't know if we'd notice this in the real world either. What would the evidence be? The highlighting would be a little off, or a code lens would be wrong, or something like that. But that happens regularly enough anyway, so much that we all know we can move the cursor, or add a line, or save the buffer, and things will eventually get back to normal.

And third, it's possible there's a bug in lsp4j, a deadlock or something. But, I think we'd notice this in the real world. LSP would just stop working periodically, which isn't a problem I've heard reported.


Thinking about the stack dump a bit more, I don't have a full picture yet. My _guess_ is that these stack lines are lsp4j waiting for the next request from the client.

```
	at java.io.BufferedInputStream.read(java.base@11.0.14/BufferedInputStream.java:271)
	- locked <0x0000000780054ce8> (a java.io.BufferedInputStream)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:79)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:113)
```

If it got there, perhaps it successfully wrote the message to `System/out`. That's a big guess though. It depends on whether the "wait for messages from client" thread is the same as the "send messages to client" thread. If they're the same, we can conclude lsp4j wrote the message. If not, we still don't know if the message was written. And even if it was written, we don't know if it was flushed.

I'm not sure what the other lsp4j stack lines are. It looks like we [call](https://github.com/clojure-lsp/clojure-lsp/blob/1d6e554ec0afd3a04c113e513f6b6724bbe0f709/cli/src/clojure_lsp/server.clj#L245) `(.startListening launcher)`, that returns a future of some sort, and [we deref it](https://github.com/clojure-lsp/clojure-lsp/blob/1d6e554ec0afd3a04c113e513f6b6724bbe0f709/cli/src/clojure_lsp/main.clj#L140). My guess is that the future resolves when the [server gets shutdown](https://github.com/clojure-lsp/clojure-lsp/blob/1d6e554ec0afd3a04c113e513f6b6724bbe0f709/cli/src/clojure_lsp/server.clj#L123), and that lets the process close. That's more theory than evidence, so take it with a grain of salt.

**Comment by mainej** [comment]
I found [where](https://github.com/eclipse/lsp4j/blob/5605ca82252833777dbef82ca13a403dbfb9d906/org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/json/StreamMessageConsumer.java#L66-L70) lsp4j writes messages (I think). It looks like it `flush`es correctly. So, that means I don't think we'd benefit from using an unbuffered stream in tests. That's too bad, because I didn't have any other ideas of things we could change on our side. 🤷

Maybe the thing to do is create a branch with `tee-system-out` enabled, and start running the integration tests against it. Then we'd get to see the actual messages (maybe).

**Comment by ericdallo** [comment]
@mainej yeah, I don't think the issue is some flush somewhere, I still think client loop is wrong somehow, but you probably spent more time than me with that.
Also, I don't see our client that different from a editor, a LSP communication should happen via a process's stdin/stdout, which is exactly what we do, maybe the client would be quite similar to editor if we remove the loop and implement some kind of listener for events that is not a loop.
